ext/dom/xpath_callbacks.c

```diff
@@ -0,0 +1,504 @@
+		xmlXPathObjectPtr obj = valuePop(ctxt);
+		ZEND_ASSERT(obj != NULL);
+		switch (obj->type) {
+			case XPATH_STRING:
+				ZVAL_STRING(param, (char *)obj->stringval);
+				break;
+			case XPATH_BOOLEAN:
+				ZVAL_BOOL(param, obj->boolval);
+				break;
+			case XPATH_NUMBER:
+				ZVAL_DOUBLE(param, obj->floatval);
+				break;
+			case XPATH_NODESET:
+				if (evaluation_mode == PHP_DOM_XPATH_EVALUATE_NODESET_TO_STRING) {
+					char *str = (char *)xmlXPathCastToString(obj);
+					ZVAL_STRING(param, str);
+					xmlFree(str);
+				} else if (evaluation_mode == PHP_DOM_XPATH_EVALUATE_NODESET_TO_NODESET) {
+					if (obj->nodesetval && obj->nodesetval->nodeNr > 0) {
+						array_init_size(param, obj->nodesetval->nodeNr);
+						zend_hash_real_init_packed(Z_ARRVAL_P(param));
+						for (int j = 0; j < obj->nodesetval->nodeNr; j++) {
+							xmlNodePtr node = obj->nodesetval->nodeTab[j];
+							zval child;
+							if (UNEXPECTED(node->type == XML_NAMESPACE_DECL)) {
+								xmlNodePtr nsparent = node->_private;
+								xmlNsPtr original = (xmlNsPtr) node;
+
+								/* Make sure parent dom object exists, so we can take an extra reference. */
+								zval parent_zval; /* don't destroy me, my lifetime is transfered to the fake namespace decl */
+								php_dom_create_object(nsparent, &parent_zval, intern);
+								dom_object *parent_intern = Z_DOMOBJ_P(&parent_zval);
+
+								php_dom_create_fake_namespace_decl(nsparent, original, &child, parent_intern);
+							} else {
+								proxy_factory(node, &child, intern, ctxt);
+							}
+							zend_hash_next_index_insert_new(Z_ARRVAL_P(param), &child);
+						}
+					} else {
+						ZVAL_EMPTY_ARRAY(param);
+					}
+				}
+				break;
+			default:
+				ZVAL_STRING(param, (char *)xmlXPathCastToString(obj));
+				break;
+		}
+		xmlXPathFreeObject(obj);
+	}
+
+	return params;
+}
+
+static void php_dom_xpath_callback_cleanup_args(zval *params, uint32_t param_count)
+{
+	if (params) {
+		for (uint32_t i = 0; i < param_count; i++) {
+			zval_ptr_dtor(&params[i]);
+		}
+		efree(params);
+	}
+}
+
+static zend_result php_dom_xpath_callback_dispatch(php_dom_xpath_callbacks *xpath_callbacks, php_dom_xpath_callback_ns *ns, xmlXPathParserContextPtr ctxt, zval *params, uint32_t param_count, const char *function_name, size_t function_name_length)
+{
+    zval callback_retval;
+
+	if (UNEXPECTED(ns == NULL)) {
+		zend_throw_error(NULL, "No callbacks were registered");
+		return FAILURE;
+	}
+
+	if (ns->mode == PHP_DOM_REG_FUNC_MODE_ALL) {
+		zend_fcall_info fci;
+		fci.size = sizeof(fci);
+		fci.object = NULL;
+		fci.retval = &callback_retval;
+		fci.param_count = param_count;
+		fci.params = params;
+		fci.named_params = NULL;
+		ZVAL_STRINGL(&fci.function_name, function_name, function_name_length);
+
+		zend_result result = zend_call_function(&fci, NULL);
+		zend_string_release_ex(Z_STR(fci.function_name), false);
+		if (UNEXPECTED(result == FAILURE)) {
+			return FAILURE;
+		}
```

Comment:
This looks wrong.

I wouldn't expect result to ever be ``FAILURE`` see the comment in ``zend_API.h``:
```
/* Can only return FAILURE if EG(active) is false during late engine shutdown.
 * If the call or call setup throws, EG(exception) will be set and the retval
 * will be UNDEF. Otherwise, the retval will be a non-UNDEF value. */
```

Reply:
Good catch, a bit surprising though.